### PR TITLE
[java] FIXED: Java 8 parsing problem with annotations for wildcards

### LINKED ISSUE #414: [java] Java 8 parsing problem with annotations for wildcards
pmd-java:
- Annotations are also allowed for generic wildcards (like NonNull-annotations in Lists)

run test case:
`net.sourceforge.pmd.lang.java.ast.ParserCornersTest.testParsersCases18()`
with
class ParserCornerCases18 modified:
`public List<@AnnotatedUsage ?> testWildCardWithAnnotation() {
    	return null;
    }`
result: parsing error
fix for java.jjt:

	void TypeArgument():
	{}
	{
	   (Annotation() {checkForBadTypeAnnotations();})* (ReferenceType() | "?" [ WildcardBounds() ])
	}

### LINKED ISSUE #415: [java] Parsing Error when having an Annotated Inner class
having a valid Java-Class:

	public class X {
		static class StaticInner {
	    		static class StaticInnerInner {
	
	        }
	    }
	
	    public void testAnnotationOnInnerClass() {
	    	final StaticInner.@Nullable StaticInnerInner x = null;
	    }
	}

Then the `@Nullable` will not be parsable - even it is the only way to annotate the inner class here.

I think the problem is somewhere in Java.jjt
`void ClassOrInterfaceType():`

### LINKED ISSUE #417: [java] Parsing Problem with Annotation for Array Member Types
following java 8 method leads to parsing error

	public Object @Nullable [] testAnnotationsToArrayElements() {
		return null;
	}

PR will follow

### PR DESCRIPTION
**Pull request by Snap252**
- added testcase to reproduce the errors
 - fix #414 
 - fix #415 
 - fix #417

### CODE REVIEW

File: pmd-java/etc/grammar/Java.jjt [hunk 1]
```diff
@@ -1488,7 +1488,7 @@ void MethodDeclaration(int modifiers) :
 {jjtThis.setModifiers(modifiers);}
 {
   [ TypeParameters() ]
-  ResultType() MethodDeclarator() [ "throws" NameList() ]
+  [(Annotation() {checkForBadTypeAnnotations();})+] ResultType() MethodDeclarator() [ "throws" NameList() ]
```
Comment (jsotuyod) [inline pmd-java/etc/grammar/Java.jjt]: I'm not sure what scenario you are trying to cover here...

File: pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java [hunk 2]
```diff
@@ -165,6 +165,27 @@ public void lambdaWithPropertyAssignment() {
...
+    
+	public Object @Nullable [] testAnnotationsToArrayElements() {
+		return null;
+	}
+	
+	private byte @Nullable [] getBytes(){
+		return null;
+	}
+	
+	public static <T extends @NonNull Enum<T>> getEnum() {
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java]: this declaration is missing the return type, this should fail to parse... you probably meant to write: `public static <T extends @NonNull Enum<T>> T getEnum() {`

File: pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java [hunk 3]
```diff
@@ -165,6 +165,27 @@ public void lambdaWithPropertyAssignment() {
...
+		return null;
+	}
+	
+	public void createNonNullArray() {
+		return new MetaColumn @NonNull [value];
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java]: your grammar update supports it, but I'd love to see a multi-dimensional array here, each dimension with it's own annotation

Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java]: Also, this method wont' compile. `void` method returning an rray

File: pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java [hunk 4]
```diff
@@ -177,12 +177,17 @@ public void lambdaWithPropertyAssignment() {
...
 	}
 	
-	public static <T extends @NonNull Enum<T>> getEnum() {
+	public static <T extends @NonNull Enum<T>> T getEnum() {
 		return null;
 	}
 	
-	public void createNonNullArray() {
-		return new MetaColumn @NonNull [value];
+	public Object createNonNullArray() {
```
Comment (jsotuyod) [inline pmd-java/src/test/resources/net/sourceforge/pmd/ast/ParserCornerCases18.java]: `public Object[] createNonNullArray() {` maybe?

**Comment by Snap252** [comment]
sorry, didn't know all further commits will end up here.... :-(

**Comment by jsotuyod** [comment]
@Snap252 don't worry. All commits to your branch are added to the PR, github assumes you are using feature branches.

**Comment by Snap252** [comment]
@jsotuyod :+1: 
When I try to close this PR, then no more changes will be imported - right?

Looking back - what would have been the right way to use git here? 
make a fork -> then create a feature branch -> commit to master -> apply them to the feature branch -> PR to the feature branch?

Would you commit these changes pull the changes in current state or is there anything I should change?

**Comment by jsotuyod** [comment]
@Snap252 I'd have created a feature branch from master, and commit there directly the test + fix for a scenario, and submitted a PR. If you wanted to work on a second issue in the meantime, I'd have created a new feature branch from master (assuming the changes are independent, even if on the same files), or from the previous feature branch if dependent (in this second case, PRs would have to be submitted and merged in order).

Please, notice there is still a question about a particular change to the grammar you haven't answered. Based on your answer I may take this as is, or suggest some changes.

File: pmd-java/etc/grammar/Java.jjt (hunk 1 above)
Comment (Snap252) [inline pmd-java/etc/grammar/Java.jjt]: stupid me ;)
... it seems I committed too fast - I 'reverted' this change

**Comment by Snap252** [comment]
I took a look at all occurences of `ClassOrInterfaceType()` in the jjt and it seems that (nearly) each of them is prefixed by a ` (Annotation() {checkForBadTypeAnnotations();})*`. I suppose they belong to each other and can be inlined. But to be sure there should be some negative-tests for java 7 and below.


Maybe this one can be used to double-check the grammar:

[https://github.com/antlr/grammars-v4/tree/master/java8](https://github.com/antlr/grammars-v4/tree/master/java8)

File: pmd-java/etc/grammar/Java.jjt (hunk 1 above)
Comment (Snap252) [inline pmd-java/etc/grammar/Java.jjt]: its me again - just did not have a test case for that scenario - now it is committed

**Comment by jsotuyod** [comment]
Merged! This will be part of PMD 5.8.0!

**Pull request merged**